subprojects/internal-testing/src/main/groovy/org/gradle/test/fixtures/file/TestFile.java

```diff
@@ -602,7 +602,7 @@ public TestFile assertIsEmptyDir() {
 
     private void visit(Set<String> names, String prefix, File file) {
         for (File child : file.listFiles()) {
-            if (child.isFile()) {
+            if (child.isFile() && child.canRead()) {
```

Comment:
test would list the file in `assertHasDescendants` otherwise

Reply:
I'd argue that an unreadable file _is_ still a descendant of its parent, it just can't be read. My understanding is that `assertHasDescendants()` checks whether there are any entries in a directory, it doesn't judge whether or not they are readable.

Reply:
it can't list and will throw an NPE at `for (File child : file.listFiles())`

Reply:
Reverted